### PR TITLE
[Spark] Support external DSV2 catalog in Restore & Clone command

### PR DESCRIPTION
**Pull request by gengliangwang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

 
Support external DSV2 catalog in RESTORE & Clone command. After the changes, the restore command supports tables from external DSV2 catalog.
For example, with
```
spark.sql.catalog.customer_catalog=org.apache.spark.sql.CustomerCatalog
```
We can query
```
SET CATALOG customer_catalog;
RESTORE TABLE t1 VERSION AS OF 0
CREATE TABLE spark.default.t2 SHALLOW CLONE t1 VERSION AS OF 1
```
Or simply
```
RESTORE TABLE customer_catalog.default.t1 VERSION AS OF 0
CREATE TABLE spark.default.t2 SHALLOW CLONE customer_catalog.default.t1 VERSION AS OF 1
```
-->

## How was this patch tested?

1. new end-to-end tests
2. new parser test cases

## Does this PR introduce _any_ user-facing changes?

Yes, users can use RESTORE & Clone command on the tables of their external DSV2 catalogs.

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @ryan-johnson-databricks @LukasRupprecht

File: spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      // To make it simple, here we insert data directly into the table path.
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (0)")
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (1)")
+      // Test 3-part identifier when the current catalog is the default catalog
+      sql(f"CREATE TABLE $destTable1 SHALLOW CLONE dummy.default.$srcTable VERSION AS OF 1")
+      checkAnswer(spark.table(destTable1), spark.range(1).toDF())
+
+      sql("SET CATALOG dummy")
+      // Test simple shallow clone command under the dummy catalog
+      sql(f"CREATE TABLE $destTable2 SHALLOW CLONE $srcTable")
```
Comment (gengliangwang) [inline spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala:92]: Note: currently the target table to be created can be only from the session catalog.
https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala#L679

We should discuss how to fix this if the target is a table from an external DSV2 catalog.

cc @ryan-johnson-databricks

File: spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala [hunk 2]
```diff
@@ -284,7 +284,8 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
     // could revert back to that.
     val sourceRelation = new UnresolvedRelation(visitMultipartIdentifier(ctx.source))
     val maybeTimeTravelSource = maybeTimeTravelChild(ctx.clause, sourceRelation)
-    val targetRelation = UnresolvedRelation(target)
```
Comment (gengliangwang) [inline spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:287]: I will follow up and  fix this in OSS Spark
```
object UnresolvedRelation {
  def apply(tableIdentifier: TableIdentifier): UnresolvedRelation =
    UnresolvedRelation(tableIdentifier.database.toSeq :+ tableIdentifier.table)
}
```

File: spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala (hunk 1 above)
Comment (ryan-johnson-databricks) [reply spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala:92]: Why is that, sorry? Is this just more legacy V1 code that can be removed as part of this PR? 
Or is there something more nefarious going on?

File: spark/src/main/scala/org/apache/spark/sql/delta/ResolveDeltaPathTable.scala [hunk 3]
```diff
@@ -19,33 +19,51 @@ package org.apache.spark.sql.delta
...
+
+    /*
+    case tt @ TimeTravel(u: UnresolvedRelation, _, _, _) if maybeSQLFile(u.multipartIdentifier) =>
+      resolveAsPathTable(u.multipartIdentifier) match {
+        case Some(r: ResolvedTable) =>
+          val relation = DataSourceV2Relation.create(r.table, Some(r.catalog), Some(r.identifier))
+          tt.copy(relation = relation)
+        case None => tt
       }
+     */
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply spark/src/main/scala/org/apache/spark/sql/delta/ResolveDeltaPathTable.scala]: Thanks, removed

File: spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTimeTravel.scala [hunk 4]
```diff
@@ -68,33 +67,23 @@ case class PreprocessTimeTravel(sparkSession: SparkSession) extends Rule[Logical
...
-          throw DeltaErrors.notADeltaTableException("RESTORE")
+      ur: UnresolvedRelation): LogicalPlan = {
+    val relation = EliminateSubqueryAliases(sparkSession.sessionState.analyzer.ResolveRelations(ur))
+    relation match {
+      case _: View =>
+        // If table exists and not found to be a view, throw not supported error
+        throw DeltaErrors.notADeltaTableException("RESTORE")
+      case tableRelation =>
+        if (tableRelation.resolved) {
+          tableRelation
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTimeTravel.scala]: Done

File: spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTimeTravel.scala [hunk 5]
```diff
@@ -68,33 +67,23 @@ case class PreprocessTimeTravel(sparkSession: SparkSession) extends Rule[Logical
...
-            catalog.getTableMetadata(tableId).tableType == CatalogTableType.VIEW) ||
-          catalog.isTempView(ur.multipartIdentifier)) {
-          // If table exists and not found to be a view, throw not supported error
-          throw DeltaErrors.notADeltaTableException("RESTORE")
+      ur: UnresolvedRelation): LogicalPlan = {
+    val relation = EliminateSubqueryAliases(sparkSession.sessionState.analyzer.ResolveRelations(ur))
+    relation match {
+      case _: View =>
+        // If table exists and not found to be a view, throw not supported error
+        throw DeltaErrors.notADeltaTableException("RESTORE")
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply spark/src/main/scala/org/apache/spark/sql/delta/PreprocessTimeTravel.scala]: This is from original code. Now it is updated.

File: spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala (hunk 1 above)
Comment (gengliangwang) [reply spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala:92]:
```
        val finalTarget = new Path(statement.targetLocation.getOrElse(
          session.sessionState.catalog.defaultTablePath(tblIdent).toString))
```
To create a table with a clone table, it needs to have a target path. But if the catalog is an external catalog plugin, there is not such an API as `defaultTablePath`.  We can create a table and get the table path. However it is tricky to do the table creation during analysis.

File: spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala (hunk 2 above)
Comment (gengliangwang) [reply spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:287]: This should be fixed on Spark 4.0: https://github.com/apache/spark/pull/42964

**Pull request closed without merging**